src/core/address_payload_unification_design.md

```diff
@@ -0,0 +1,556 @@
+
+  graph() { return this._graph; }
+  address() { return this._address; }
+
+  get() {
+    this._ensureUpToDate();
+    if (this._index == null) {
+      return undefined;
+    } else {
+      return this._graph._nodes[this._index].value;
+    }
+  }
+
+  neighbors(options?: NeighborhoodOptions) {
+    this._ensureUpToDate();
+    if (this._index == null) {
+      return [].values();
+    } else {
+      return this._graph._internalNeighbors(this._index, options);
+    }
+  }
+
+  _ensureUpToDate() {
+    if (this._graph._indexVersion !== this._indexVersion) {
+      this._indexVersion = this._graph._indexVersion;
+      this._index = this._graph._addressToIndex(this._address);
+    }
+  }
+}
+```
+
+The salient point of this implementation is that it enables fast access
+of graph operations by remaining inside the integer-indexed layer,
+without leaking to clients the details of how this works. Notable in
+this implementation is the `_indexVersion` field, which enables a graph
+to garbage-collect, compress, permute, or otherwise change its node
+index without breaking extant references and without needing to keep
+track of them: any such operation simply increments the graph’s
+`_indexVersion` attribute, and any `IndexedNodeReference` will pick up
+the change on any subsequent actions. Again, though, this is just a
+convenient implementation detail.
+
+Clients won’t directly use `IndexedNodeReference`, though; this will be
+a non-exported class of the `core/graph.js` module. Instead, they will
+provide a semantic extension of `NodeDelegateReference`:
+
+```javascript
+class NodeDelegateReference implements NodeReference {
+  _base: NodeReference;
+  constructor(base: NodeReference) {
+    this._base = base;
+  }
+  graph() { return this._base.graph(); }
+  address() { return this._base.address(); }
+  get() { return this._base.get(); }
+  neighbors(options?: NeighborhoodOptions) {
+    return this._base.neighbors(options);
+  }
+}
+```
+
+(Note: `NodeDelegateReference` should actually have `_base` as a
+`Symbol`, but doing so requires Flow-appeasing hackery that I don’t want
+to include in this high-level overview.)
+
+By extending this class, plugins gain great power:
+
+```javascript
+class GithubNodeReference extends NodeDelegateReference {
+  constructor(base: NodeReference) {
+    super(base);
+    const address = base.address();
+    if (address.pluginName !== "sourcecred/github-beta") {
+      throw new Error("pluginName: " + address.pluginName);
+    }
+  }
+}
+
+class PullRequestReference extends GithubNodeReference {
+  constructor(base: NodeReference) {
+    super(base);
+  }
+  *comments(): Iterator<PullRequestCommentReference> {
+    for (const neighbor of this.neighbors({
+      direction: "OUT",
+      nodeType: "PULL_REQUEST_COMMENT",
+      edgeType: "CONTAINS",
+    })) {
+      if (!(neighbor instanceof PullRequestCommentReference)) {
+        throw new Error(neighbor.constructor.name);
```

Comment:
Is there any reason not to move this pattern down into the `neighbors` method? I don't look forward to re-implementing this for every `.*Reference` accessor

Reply:
We could add e.g. a `ReferenceConstructorName` option to `NeighborOptions`, which either filters out references that don't have the right constructor, or requires that every neighbor have that constructor and throw an error if not (ie: the same semantics here).

My point is, your including this in the generic example for using a reference suggests that we expect basically every caller to neighbors to apply this logic, in which case we not be expecting clients to all re-write this boilerplate over and over.